enhancements/windows-containers/health-management.md

```diff
@@ -0,0 +1,311 @@
+
+When the WICD cleanup command is executed by WMCO, all services listed in the ConfigMap will be stopped and removed.
+It will then delete the instance's Node object.
+
+### Test Plan
+
+Most functionality described within this enhancement is already tested as part of the extensive end-to-end
+tests in the WMCO repo. As the different commands of WICD are developed, they can be integrated into WMCO, and
+tested through the existing tests. New tests will be added to test functionality that wasn't already present, such
+as validating the state of the services ConfigMap, and ensuring that Windows service state is reverted after it
+is tampered with.
+
+### Graduation Criteria
+
+This enhancement is targeted for OpenShift 4.11/ WMCO version 6.0.0.
+
+#### Dev Preview -> Tech Preview
+
+The features added here can be previewed in the Community 5.y.z release.
+
+#### Tech Preview -> GA
+
+As the functionality described in this enhancement will be integrated into WMCO, the normal release process
+of WMCO will be followed. That being a community operator release for early preview, followed by the official Red Hat
+operator release.
+
+#### Removing a deprecated feature
+
+N/A, as the changes made in the enhancement should be invisible to the user.
+
+### Upgrade / Downgrade Strategy
+
+When WMCO is upgraded, it will create the new Windows service ConfigMap, and then for each existing Windows Node:
+1) Stop the WICD service
+2) Run the WICD `cleanup` command with the `--upgrade` flag to stop and remove all services without removing the Node
+   object.
+2) Copy over all new and modified files in the payload, including WICD.
+3) Update the desired version annotation on the Node, to the current WMCO version.
+4) Start the WICD service.
+
+WICD, triggered by a mismatch between desired node version and current node version labels will correct the state
+by using the ConfigMap of the version it is upgrading to, in order to configure the instance normally. As part of this
+process, the Node version label is patched from the previous version to the desired version.
+
+There should be no difference in the upgrade process between Machine and BYOH instances.
+
+Downgrades are not supported through OLM, and thus are not supported by WMCO.
+
+### Version Skew Strategy
+
+This enhancement has the potential to reduce issues related to having Windows Nodes which were configured using an
+older, now uninstalled WMCO version, on a more recent OpenShift cluster. When a compatible WMCO version is installed,
+the services ConfigMap provides a way for WICD to upgrade a Node cleanly from one version to another. This is true when
+upgrading only from WMCO 5.0.0 onwards, as WICD will have no way to cleanly upgrade without either a service ConfigMap
+from the previous version, or the `openshift-managed` label on the Windows services themselves.
+
+Any Windows Nodes from before 5.0.0 should have been deconfigured with the associated WMCO version. At this point, the
```

Comment:
This is a breaking change and is not acceptable. Why can't WMCO do the deconfiguration as it is doing today for this version? We can remove the WMCO deconfiguration in 6.0.0.

Reply:
I misworded this. I meant going into 6.0.0 any Nodes from before 5.0.0 should have been deconfigured. I will reword it.